pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -18,88 +18,128 @@ import (
 )
 
 const (
-	defaultTenantLabel         = "__tenant_id__"
-	retainExistingPrefix       = "original_"
-	originalDefaultTenantLabel = retainExistingPrefix + defaultTenantLabel
-	maxConcurrency             = 16
+	defaultTenantLabel   = "__tenant_id__"
+	retainExistingPrefix = "original_"
+	maxConcurrency       = 16
 )
 
 // NewQueryable returns a queryable that iterates through all the tenant IDs
 // that are part of the request and aggregates the results from each tenant's
 // Querier by sending of subsequent requests.
-// The result contains a label tenantLabelName to identify the tenant ID that
+// The result contains a label "__tenant_id__" to identify the tenant ID that
 // it originally resulted from.
-// If the label tenantLabelName is already existing, its value is overwritten
+// If the label "__tenant_id__" is already existing, its value is overwritten
 // by the tenant ID and the previous value is exposed through a new label
-// prefixed with "original_". This behaviour is not implemented recursively
+// prefixed with "original_". This behaviour is not implemented recursively.
 func NewQueryable(upstream storage.Queryable) storage.Queryable {
-	return &mergeQueryable{
-		upstream: upstream,
+	return NewGenericQueryable(defaultTenantLabel, tenantQuerierCallback(upstream))
+}
+
+func tenantQuerierCallback(queryable storage.Queryable) GenericQuerierCallback {
+	return func(ctx context.Context, mint int64, maxt int64) ([]string, []storage.Querier, bool, error) {
+		tenantIDs, err := tenant.TenantIDs(ctx)
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		var queriers = make([]storage.Querier, len(tenantIDs))
+		for pos, tenantID := range tenantIDs {
+			q, err := queryable.Querier(
+				user.InjectOrgID(ctx, tenantID),
+				mint,
+				maxt,
+			)
+			if err != nil {
+				return nil, nil, false, err
+			}
+			queriers[pos] = q
+		}
+
+		// ensure to byPass a single querier and avoid adding a tenant label to
+		// ensure backward compatabiltiy when enabling multi-tenant query
+		// federation.
+		byPassWithSingleQuerier := true
+
+		return tenantIDs, queriers, byPassWithSingleQuerier, nil
+	}
+}
+
+// GenericQuerierCallback returns the underlying queriers and labelValues
+// relevant for the query. If the callback returns exactly a single querier,
+// byPassWithSingleQuerierUsing determines if it should still be using the
+// mergeQuerier (and adding a `identifyingLabelName` label) or if just the
+// underlying querier should be returned.
+type GenericQuerierCallback func(ctx context.Context, mint int64, maxt int64) (labelValues []string, queriers []storage.Querier, byPassWithSingleQuerier bool, err error)
```

Comment:
In what scenario does it make sense for callback to return `byPassWithSingleQuerier=false`?

Reply:
I think in general the callback should be returning `byPassWithSingleQuerier=false`. Even though we only have a single cluster we are querying from, we still want the `idLabelName` on the results. 

In my use case I want to able to aggregate results from multiple cluster and I want to make sure even if there's only a single cluster returned by the call that this cluster is identifiable using the `idLabelName`. Once a user adds a second cluster later on, there is no user visible change to the results if we also have the label on the single querier case. 

The reason why for tenant federation `byPassWithSingleQuerier=true`, is so that we don't have a breaking behavior, when enabling tenant federation in a cortex cluster.

Reply:
> Even though we only have a single cluster we are querying from, we still want the `idLabelName` on the results.

Makes sense to me.

> The reason why for tenant federation `byPassWithSingleQuerier=true`, is so that we don't have a breaking behavior, when enabling tenant federation in a cortex cluster.

This feature is still marked as experimental. As long as we properly document the change in the changelog, it's ok to break current behavior, esp. if it is not quite correct. And supporting only single case in the code will be easier. WDYT?

Reply:
I think I am less worried about a breaking change in the experimental feature, than the future experience of turning tenant federation on. With having an additonal "id" label in the single user queries, it would be quite disruptive, while otherwise it would be not changing anything. I still think its worthy of keeping. 

//cc @jtlisi wdyt?

Reply:
So just to clarify the breaking change here would be always returning the `__tenant_id__` on every single query?

I'm personally fine with that change. I honestly don't see it being a change that bothers people. I also think it would help end users become more familiar with tenancy in Cortex.

Reply:
Exactly that, once tenant federation is switched on, every query would return a `__tenant_id__` label.

OK maybe I was a bit too worried how user facing change it is. I will change this PR accordingly and add a `[CHANGE]` note to the changelog.

Reply:
Discussed offline. We agreed on keeping the mechanism to bypass mergeQuerier logic when only single tenant is queried.

Reply:
I have moved `byPassWithSingleQuerier` to now be a parameter to `NewQueryable` and added a test case

Reply:
> I have moved byPassWithSingleQuerier to now be a parameter to NewQueryable and added a test case

Awesome, I much prefer doing it in the creation function and not the callback